### PR TITLE
My Jetpack Add filter in my jetpack activation

### PR DESCRIPTION
**Pull request by leogermani**
#### Changes proposed in this Pull Request:

* Added a filter to allow stand-alone plugins to add product specific activation routines

#### Jetpack product discussion

p9dueE-55u-p2

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Test the activation of different products with different combination of plugins installed/active. All should work the same way as before

Testing the hook.

In the Search plugin, add a callback:

```
add_filter( 'my_jetpack_search_activation', function( $result ) {
   if ( is_wp_error( $result ) return $result;
  // do things
  return true;
} );
```
And verify your coded runs when you activate the product via My Jetpack

### CODE REVIEW

File: projects/packages/my-jetpack/src/products/class-hybrid-product.php [hunk 1]
```diff
@@ -67,11 +67,10 @@ public static function activate_plugin() {
...
+	 * @param bool|WP_Error $product_activation Is the result of the top level activation actions. You probably won't do anything if it is an WP_Error.
+	 * @return bool|WP_Error
 	 */
-	public static function activate() {
-
-		$product_activation = parent::activate();
+	public static function do_product_specific_activation( $product_activation ) {
 
 		if ( is_wp_error( $product_activation ) ) {
 			return $product_activation;
```
Comment (kangzj) [inline projects/packages/my-jetpack/src/products/class-hybrid-product.php:76]: Not related to the PR but I suspect the `Jetpack` class reference would need a namespace on L81.

**Review by kangzj — DISMISSED** [review]
The PR works as expected and looks good to me 👍 

I found two issues by the way testing, not related to the PR tho. One of which is left by a comment below. The other one is that, Search status is not correctly reflected, code:

https://github.com/Automattic/jetpack/blob/c740f05c785560299b7d834655c034dcec37d906/projects/packages/my-jetpack/src/products/class-search.php#L198

Where I suppose we need to test whether search module is active as well possibly by doing ` (new Modules)->is_active( static::$module_name )` cuz `Jetpack` plugin might not be available.

File: projects/packages/my-jetpack/src/products/class-hybrid-product.php (hunk 1 above)
Comment (kangzj) [reply projects/packages/my-jetpack/src/products/class-hybrid-product.php:76]: We possibly need to use `Modules` class on L81 as well.

Comment (kangzj) [reply projects/packages/my-jetpack/src/products/class-hybrid-product.php:76]: We possibly need to use `Modules` class on L81 as well.

Comment (leogermani) [reply projects/packages/my-jetpack/src/products/class-hybrid-product.php:76]: > Not related to the PR but I suspect the Jetpack class reference would need a namespace on L81.

Yes, I noticed that while working on #24335 and fixed it there (these 2 PRs will conflict as soon as one of them gets merged, but that's ok, I'm expecting it)

**Comment by leogermani** [comment]
> Search status is not correctly reflected

@kangzj It seems we will also have to add a filter to `is_active`? Or using the `Modules` class would cover everything (including activation)?

**Comment by kangzj** [comment]
> @kangzj It seems we will also have to add a filter to is_active? Or using the Modules class would cover everything (including activation)?

IMO `Modules` would cover both `is_active` and `activation` well (I didn't  test activation, if it doesn't work I'll fix it - looks promising) - AFAIK WordAds and a couple other standalones would follow the same approach as Search, so it would be great if we could support it from My Jetpack.

**Comment by leogermani** [comment]
@kangzj I just looked at the Modules Class in the Status package... it's new and I was not aware that this has been extracted from the Jetpack plugin!

That's awesome! I'll adapt this PR to use the Modules class then.

I'll leave the filter as it might be useful, but maybe we won't need it!

**Comment by leogermani** [comment]
Actually, activating a module with the `Modules` class will only work when Jetpack plugin is active... it doesn't serve our purpose here...

Discussion: p1652392614473149-slack-CBG1CP4EN

**Pull request merged**